src/main/kotlin/com/tradeshift/blayze/Model.kt

```diff
@@ -46,19 +44,58 @@ import com.tradeshift.blayze.features.Text
  * @param textFeatures          The text features used by this model.
  * @param categoricalFeatures   The categorical features used by this model.
  * @param gaussianFeatures      The gaussian features used by this model.
- * @param pseudoCount           Pseudo count of observed outcomes. Is added to all outcome counts in the [priorCounts].
+ * @param priorPseudoCount      Pseudo count of observed outcomes. Is added to all outcome counts in the [priorCounts].
  */
 class Model(
-        private val priorCounts: Map<Outcome, Int> = mapOf(),
-        private val textFeatures: Map<FeatureName, Text> = mapOf(),
-        private val categoricalFeatures: Map<FeatureName, Categorical> = mapOf(),
-        private val gaussianFeatures: Map<FeatureName, Gaussian> = mapOf(),
-        private val pseudoCount: Int = 0
+        val priorCounts: Map<Outcome, Int> = mapOf(),
+        val textFeatures: Map<FeatureName, Text> = mapOf(),
+        val categoricalFeatures: Map<FeatureName, Categorical> = mapOf(),
+        val gaussianFeatures: Map<FeatureName, Gaussian> = mapOf(),
+        val priorPseudoCount: Int = 0
 ) {
 
+    /**
+     * Parameters of a model.
+     *
+     * @param priorPseudoCount Pseudo count of observed outcomes. Is added to all outcome counts in the [priorCounts].
+     * @param text parameters of the text features. See [Multinomial.Parameters].
+     * @param categorical parameters of the categorical features. See [Multinomial.Parameters].
+     * @param gaussian parameters of the gaussian features. See [Gaussian.Parameters].
+     */
+    data class Parameters(
+            val priorPseudoCount: Int = 0,
+            val text: Map<FeatureName, Multinomial.Parameters> = mapOf(),
+            val categorical: Map<FeatureName, Multinomial.Parameters> = mapOf(),
+            val gaussian: Map<FeatureName, Gaussian.Parameters> = mapOf()
+    )
+
+    // Just a tiny optimization to cache the default logPrior
+    private val defaultLogPrior: Map<String, Double> by lazy {
+        logPrior(priorPseudoCount)
+    }
+
+    /**
+     * Return a new model with updated default [Parameters].
+     */
+    fun withParameters(parameters: Parameters): Model {
+        val text = withParameters(textFeatures, parameters.text, { Text() })
+        val categorical = withParameters(categoricalFeatures, parameters.categorical, { Categorical() })
+        val gaussian = withParameters(gaussianFeatures, parameters.gaussian, { Gaussian() })
+        return Model(priorCounts, text, categorical, gaussian, parameters.priorPseudoCount)
+    }
+
+    private fun <V, P, F : Feature<F, V, P>> withParameters(features: Map<FeatureName, F>, parameters: Map<FeatureName, P>, creator: () -> F ): Map<FeatureName, F> {
+        var cf = features
+        for ((n, p) in parameters) {
+            val f = cf[n] ?: creator()
+            cf = cf.plus(n to f.withParameters(p))
```

Comment:
Consider just making cf mutable here instead of copying it multiple times.
It is probably OK here since the feature implementations aren't copying the large data maps in `withParameters` and there probably aren't that many different features.
But generally the kotlin immutable types are not persistent data structures like in Scala, but rather just make new copies.

Reply:
good point